File: baekjoon/rust/gcd_n_k_eq_1_11689_13926.rs
// Title: GCD(n, k) = 1
// Link: https://www.acmicpc.net/problem/13926
// Link: https://www.acmicpc.net/problem/11689

use std::collections::{BTreeMap, HashMap, HashSet};
use std::io::{stdin, stdout, BufRead, BufReader, BufWriter, Write};

macro_rules! read_to_tuple {
    ($buffer:expr, $($t:ty),+) => {{
        let mut line = String::new();
        let buffer = &mut $buffer;
        buffer.read_line(&mut line).unwrap();
        let mut it = line.trim().split_whitespace();
        (
            $(it.next().unwrap().parse::<$t>().unwrap()),+
        )
    }}
}

macro_rules! read_to_vector {
    ($buffer:expr, $t:ty) => {{
        let mut line = String::new();
        let buffer = &mut $buffer;
        buffer.read_line(&mut line).unwrap();
        line.trim()
            .split_whitespace()
            .map(|x| x.parse::<$t>().unwrap())
            .collect::<Vec<$t>>()
    }};
}

type Func_x_n = fn(u128, u128) -> u128;
type Func_n_f = fn(u128, Func_x_n) -> u128;

fn gcd(mut a: u128, mut b: u128) -> u128 {
    if a == 0 {
        return b;
    }
    return gcd(b % a, a);
}

fn f1(mut x: u128, mut n: u128) -> u128 {
    ((x * x) + 1) % n
}

fn f2(mut x: u128, mut n: u128) -> u128 {
    ((x * x) + 2) % n
}

fn f3(mut x: u128, mut n: u128) -> u128 {
    ((x * x) + 3) % n
}

fn rho(mut n: u128, mut f: Func_x_n) -> u128 {
    let mut x = 2;
    let mut y = 2;
    let mut d = 1;
    let mut count = 0;
    while d == 1 {
        x = f(x, n);
        y = f(f(y, n), n);
        if x > y {
            d = gcd(x - y, n);
        } else {
            d = gcd(y - x, n);
        }
        count += 1;
        if count > 100000 {
            return 0;
        }
    }
    if d == n {
        return 0;
    }
    return d;
}

fn get_factoring(n: u128, s: &mut HashSet<u128>, f_dict: &mut HashMap<u128, bool>) {
    f_dict.insert(n, true);

    let f = rho(n, f1);

    if n == 4 || n == 8 || n == 16 {
        s.insert(2);
        return;
    } else if n == 25 || n == 125 {
        s.insert(5);
        return;
    } else if n == 9 || n == 27 || n == 81 {
        s.insert(3);
        return;
    }

    if f == 0 {
        let f_2 = rho(n, f2);
        let f_3 = rho(n, f3);
        if ((f_2 == f_3) && (f_3 == f)) && n != 4 && n != 8 {
            s.insert(n);
        } else {
            if f_2 != 0 {
                let a = n / f_2;
                let b = f_2;
                if !f_dict.contains_key(&a) {
                    get_factoring(a, s, f_dict);
                }
                if !f_dict.contains_key(&b) {
                    get_factoring(b, s, f_dict);
                }
            }
            if f_3 != 0 {
                let a = n / f_3;
                let b = f_3;
                if !f_dict.contains_key(&a) {
                    get_factoring(a, s, f_dict);
                }
                if !f_dict.contains_key(&b) {
                    get_factoring(b, s, f_dict);
                }
            }
        }
    } else {
        if f != 0 {
            let a = n / f;
            let b = f;
            if !f_dict.contains_key(&a) {
                get_factoring(a, s, f_dict);
            }
            if !f_dict.contains_key(&b) {
                get_factoring(b, s, f_dict);
            }
        }
    }
}

fn num_gcd(n: u128) -> u128 {
    if n == 1 {
        return 1;
    }
    let mut factors = HashSet::<u128>::new();
    let mut f_dict = HashMap::<u128, bool>::new();
    get_factoring(n, &mut factors, &mut f_dict);
    let mut up = 1;
    let mut down = 1;
    for f in factors {
        up *= (f - 1);
        down *= f;
    }
    n * up / down
}

fn main() {
    let mut reader = BufReader::new(stdin());
    let mut writer = BufWriter::new(stdout());

    let n = read_to_tuple!(reader, u128);
    writeln!(writer, "{}", num_gcd(n));
    writer.flush().unwrap();
}

File: baekjoon/python/gcd_n_k_eq_1_11689_13926.py
# Title: GCD(n, k) = 1
# Link: https://www.acmicpc.net/problem/11689:
# Link: https://www.acmicpc.net/problem/13926:

import sys

sys.setrecursionlimit(10 ** 6)


def read_list_int():
    return list(map(int, sys.stdin.readline().strip().split(' ')))


def read_single_int():
    return int(sys.stdin.readline().strip())


def gcd(a, b):
    while a % b != 0:
        a, b = b, a % b
    return b


def f1(x, n): return ((x*x) + 1) % n


def f2(x, n): return ((x*x) + 2) % n


def f3(x, n): return ((x*x) + 3) % n


def rho(n, f):
    x = 2
    y = 2
    d = 1
    count = 0
    while d == 1:
        x = f(x, n)
        y = f(f(y, n), n)
        d = gcd(abs(x - y), n)
        count += 1
        if count > 100000:
            return False
    if d == n:
        return False
    return d


def get_factoring(n, s, f_dic):
    # print('get_factoring {}'.format(n))
    f_dic[n] = True

    f = rho(n, f1)

    if n == 4 or n == 8 or n == 16:
        s.add(2)
        return
    elif n == 25 or n == 125:
        s.add(5)
        return
    elif n == 9 or n == 27 or n == 81:
        s.add(3)
        return

    if not f:
        f_2 = rho(n, f2)
        f_3 = rho(n, f3)
        if f_2 == f_3 == f and n != 4 and n != 8:
            s.add(n)
        else:
            if f_2:
                a, b = n//f_2, f_2
                if a not in f_dic:
                    get_factoring(a, s, f_dic)
                if b not in f_dic:
                    get_factoring(b, s, f_dic)
            if f_3:
                a, b = n//f_3, f_3
                if a not in f_dic:
                    get_factoring(a, s, f_dic)
                if b not in f_dic:
                    get_factoring(b, s, f_dic)
    else:
        if f:
            a, b = n//f, f
            if a not in f_dic:
                get_factoring(a, s, f_dic)
            if b not in f_dic:
                get_factoring(b, s, f_dic)


def num_gcd(n):
    if n == 1:
        return 1
    factors = set()
    f_dic = {}
    get_factoring(n, factors, f_dic)
    # print(factors)
    up, down = 1, 1
    for f in factors:
        up *= (f-1)
        down *= f
    return n * up // down


def sol(n):
    r = n
    for p in range(2, int(n**0.5)+1):
        if n % p == 0:
            while n % p == 0:
                n = n // p
            r = r - r//p
    if n > 1:
        r = r - r//n
    return r


if __name__ == '__main__':
    n = read_single_int()
    print(num_gcd(n))

    # for n in range(1000000000000000000, 0, -1):
    #     print('{} : {}'.format(n, num_gcd(n)))

    # print(num_gcd(999999999999999994))

    # print(rho(1546412477693, f1))
    # print(rho(1546412477693, f2))
    # print(rho(1546412477693, f3))
